### PR TITLE
wereAddressesSpentFrom API call - WIP

### PR DESCRIPTION
**Pull request by alon-e**
an implementation of #503 
also moved file signature validation out of Snapshot

still missing a static version of prev. epochs spent addresses.

### CODE REVIEW

File: src/main/java/com/iota/iri/service/API.java [hunk 1]
```diff
@@ -100,6 +100,8 @@
     private HashMap<InetAddress,AtomicInteger> broadcastStoreCounters;
     private AtomicLong broadcastStoreTimer;
 
+    private ConcurrentHashMap<Hash, Boolean> previousEpochsSpentAddresses;
```
Comment (th0br0) [inline src/main/java/com/iota/iri/service/API.java]: As we don't modify the map after creation, how about we make this a regular immutable Map?

File: src/main/java/com/iota/iri/service/API.java [hunk 2]
```diff
@@ -355,6 +383,65 @@ private AbstractResponse process(final String requestString, InetSocketAddress s
...
 
+    private AbstractResponse wereAddressesSpentFromStatement(List<String> addressesStr) throws Exception {
+        final List<Hash> addresses = addressesStr.stream().map(Hash::new).collect(Collectors.toList());
+        final boolean[] states = new boolean[addresses.size()];
+        int index = 0;
+
+        for (Hash address : addresses) {
+            states[index++] = wasAddressSpentFrom(address);
+        }
+        return GetInclusionStatesResponse.create(states);
```
Comment (th0br0) [inline src/main/java/com/iota/iri/service/API.java]: Rename the response model?

File: src/main/java/com/iota/iri/service/API.java [hunk 3]
```diff
@@ -355,6 +383,65 @@ private AbstractResponse process(final String requestString, InetSocketAddress s
...
+    }
+
+    private boolean wasAddressSpentFrom(Hash address) throws Exception {
+        if (previousEpochsSpentAddresses.containsKey(address)) {
+            return true;
+        }
+        Set<Hash> hashes = AddressViewModel.load(instance.tangle, address).getHashes();
```
Comment (th0br0) [inline src/main/java/com/iota/iri/service/API.java]: Potential bottleneck for addresses with many txs (e.g. snapshot)

File: src/main/java/com/iota/iri/service/API.java [hunk 4]
```diff
@@ -355,6 +383,65 @@ private AbstractResponse process(final String requestString, InetSocketAddress s
...
+                //confirmed
+                if (tx.snapshotIndex() != 0) {
+                    return true;
+                }
+                //pending
+                Hash tail = findTail(hash);
+                if (tail == null) {
+                    return false;
+                }
+                return (BundleValidator.validate(instance.tangle, tail  ).size() != 0);
```
Comment (th0br0) [inline src/main/java/com/iota/iri/service/API.java]: formatting

File: src/main/java/com/iota/iri/service/API.java [hunk 5]
```diff
@@ -155,6 +162,24 @@ public void handleRequest(final HttpServerExchange exchange) throws Exception {
         server.start();
     }
 
+    private void readPreviousEpochsSpentAddresses() {
+        if (!SignedFiles.isFileSignatureValid("/previousEpochsSpentAddresses.txt", "/previousEpochsSpentAddresses.sig",
+                Snapshot.SNAPSHOT_PUBKEY, Snapshot.SNAPSHOT_PUBKEY_DEPTH, Snapshot.SPENT_ADDRESSES_INDEX)) {
+            log.error("Failed to load previousEpochsSpentAddresses - signature failed.");
```
Comment (GalRogozinski) [inline src/main/java/com/iota/iri/service/API.java]: are you not supposed to `return` or `throw` after this line?

File: src/main/java/com/iota/iri/service/API.java [hunk 6]
```diff
@@ -355,6 +384,64 @@ private AbstractResponse process(final String requestString, InetSocketAddress s
...
+            //spend
+            if (tx.value() < 0) {
+                //confirmed
+                if (tx.snapshotIndex() != 0) {
+                    return true;
+                }
+                //pending
+                Hash tail = findTail(hash);
+                if (tail == null) {
+                    return false;
```
Comment (GalRogozinski) [inline src/main/java/com/iota/iri/service/API.java]: Please remove

File: src/main/java/com/iota/iri/service/API.java [hunk 7]
```diff
@@ -355,6 +384,64 @@ private AbstractResponse process(final String requestString, InetSocketAddress s
...
+                }
+                return (BundleValidator.validate(instance.tangle, tail).size() != 0);
```
Comment (GalRogozinski) [inline src/main/java/com/iota/iri/service/API.java]: if this statement is `true` then return `true`.
Else continue

File: src/main/java/com/iota/iri/service/API.java [hunk 8]
```diff
@@ -355,6 +384,64 @@ private AbstractResponse process(final String requestString, InetSocketAddress s
...
+            Set<Hash> approvees = tx.getApprovers(instance.tangle).getHashes();
+            for (Hash approvee : approvees) {
+                TransactionViewModel nextTx = TransactionViewModel.fromHash(instance.tangle, approvee);
+                if (nextTx.getBundleHash().equals(bundleHash)) {
+                    tx = nextTx;
+                    break;
+                }
+            }
+            //no related approvee found
+            break;
```
Comment (GalRogozinski) [inline src/main/java/com/iota/iri/service/API.java]: so what is the point of the while?

**Pull request merged**